### PR TITLE
 Convert resource to pojo objects when used as stack outputs. 

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/pulumi/pulumi/issues/1858

Previously, a stack output that was a resource would look like:

```
vpcInstance    : "urn:pulumi:cyfg::containers-fargate::awsinfra:x:ec2:Vpc::default-vpc"
```

Now it will look like:

```
  + vpcInstance    : {
      + instance         : {
          + arn                         : "arn:aws:ec2:us-west-1:153052954103:vpc/vpc-f006bd94"
          + assignGeneratedIpv6CidrBlock: false
          + cidrBlock                   : "172.31.0.0/16"
          + defaultNetworkAclId         : "acl-df3d56bb"
          + defaultRouteTableId         : "rtb-ce255aaa"
          + defaultSecurityGroupId      : "sg-11134f76"
          + dhcpOptionsId               : "dopt-68ea3e0c"
          + enableClassiclink           : false
          + enableClassiclinkDnsSupport : false
          + enableDnsHostnames          : true
          + enableDnsSupport            : true
          + id                          : "vpc-f006bd94"
          + instanceTenancy             : "default"
          + mainRouteTableId            : "rtb-ce255aaa"
          + ownerId                     : "153052954103"
          + urn                         : "urn:pulumi:cyfg::containers-fargate::aws:ec2/vpc:Vpc::default-vpc"
        }
      + publicSubnetIds  : [
      +     [0]: "subnet-b0dbbbe8"
      +     [1]: "subnet-10025474"
        ]
      + publicSubnets    : [
      +     [0]: {
              + instance  : {
                  + arn                        : "arn:aws:ec2:us-west-1:153052954103:subnet/subnet-b0dbbbe8"
                  + assignIpv6AddressOnCreation: false
                  + availabilityZone           : "us-west-1c"
                  + availabilityZoneId         : "usw1-az3"
                  + cidrBlock                  : "172.31.0.0/20"
                  + id                         : "subnet-b0dbbbe8"
                  + mapPublicIpOnLaunch        : true
                  + ownerId                    : "153052954103"
                  + urn                        : "urn:pulumi:cyfg::containers-fargate::awsinfra:x:ec2:Vpc$aws:ec2/subnet:Subnet::public-0"
                  + vpcId                      : "vpc-f006bd94"
                  + vpc                        : "urn:pulumi:cyfg::containers-fargate::awsinfra:x:ec2:Vpc::default-vpc"
                }
              + subnetId  : "subnet-b0dbbbe8"
              + subnetName: "public-0"
              + urn       : "urn:pulumi:cyfg::containers-fargate::awsinfra:x:ec2:Vpc$awsinfra:x:ec2:Subnet::public-0"
            }
      +     [1]: {
              + instance  : {
                  + arn                        : "arn:aws:ec2:us-west-1:153052954103:subnet/subnet-10025474"
                  + assignIpv6AddressOnCreation: false
                  + availabilityZone           : "us-west-1a"
                  + availabilityZoneId         : "usw1-az1"
                  + cidrBlock                  : "172.31.16.0/20"
                  + id                         : "subnet-10025474"
                  + mapPublicIpOnLaunch        : true
                  + ownerId                    : "153052954103"
                  + urn                        : "urn:pulumi:cyfg::containers-fargate::awsinfra:x:ec2:Vpc$aws:ec2/subnet:Subnet::public-1"
                  + vpcId                      : "vpc-f006bd94"
                  + vpc                        : "urn:pulumi:cyfg::containers-fargate::awsinfra:x:ec2:Vpc::default-vpc"
                }
              + subnetId  : "subnet-10025474"
              + subnetName: "public-1"
              + urn       : "urn:pulumi:cyfg::containers-fargate::awsinfra:x:ec2:Vpc$awsinfra:x:ec2:Subnet::public-1"
            }
        ]
      + urn              : "urn:pulumi:cyfg::containers-fargate::awsinfra:x:ec2:Vpc::default-vpc"
    }
```

Large, but actually useful detailed information that is more in line with the intuition someone wants here.

### CODE REVIEW

File: sdk/nodejs/runtime/stack.ts [hunk 1]
```diff
@@ -64,9 +65,97 @@ class Stack extends ComponentResource {
...
+            return { uri: (<asset.RemoteArchive>prop).uri };
+        }
+
+        return undefined;
+    }
+
+    if (Resource.isInstance(prop)) {
+        // Emit a resource as a normal pojo.  But filter out all our internal properties so that
+        // they don't clutter the display/checkpoint with values not relevant to the application.
+        return serializeAllKeys(n => !n.startsWith("__"));
```
Comment (ellismg) [inline sdk/nodejs/runtime/stack.ts]: Should we be more scoped in what we actually filter?  Would there ever be other keys that start with "__" which do not logically belong to Pulumi?

**Comment by lukehoban** [comment]
The one thing I worry about here is that folks will now want to re-hydrate the resource from this output, especially via StackReference, but that won’t work.  Anything we can/should do about that?

**Comment by CyrusNajmabadi** [comment]
> The one thing I worry about here is that folks will now want to re-hydrate the resource from this output, especially via StackReference, but that won’t work. 

I'm not really seeing the concern here... can you expand on your concern here.  Why/how would they go about this attempt at 'rehydration'?

**Comment by CyrusNajmabadi** [comment]
Going to merge this for now.  If you think there's a large concern here, let me know.

**Comment by pgavlin** [comment]
> The one thing I worry about here is that folks will now want to re-hydrate the resource from this output, especially via StackReference, but that won’t work.

In what sense won't that work? A `get` using the resource's `id` property should do the trick.

**Comment by lukehoban** [comment]
> In what sense won't that work? A get using the resource's id property should do the trick.

Sure - but they'll think they can just export a resource from one stack and then use it from another stack.  They will assume that in a sense the whole point of stack reference is to not have to do a separate `.get` - the StackReference is the "get".

What I could vaguely imagine is that we actually find some way to rehydrate these automatically into real Resource objects as part of a `StackReference` - lots of things I don't understand about that - but seems like it's what folks will expect when they first use `StackReference` along with the changes here.

**Comment by pgavlin** [comment]
I think that's a reasonable concern. I'd suggest that we stick with what we have--it's clearly better than what was there before--and see how things shake out.

**Comment by krishnamauryaext** [comment]
> > In what sense won't that work? A get using the resource's id property should do the trick.
> 
> Sure - but they'll think they can just export a resource from one stack and then use it from another stack. They will assume that in a sense the whole point of stack reference is to not have to do a separate `.get` - the StackReference is the "get".
> 
> What I could vaguely imagine is that we actually find some way to rehydrate these automatically into real Resource objects as part of a `StackReference` - lots of things I don't understand about that - but seems like it's what folks will expect when they first use `StackReference` along with the changes here.

I am in exactly the situation that @lukehoban pointed out. I am exporting the serviceAccount from one project so the same can be used by multiple projects down the line by stack referencing the output. The value I get using the `stackReference.getOutput("")` is of type `Output<t>` whereas the it needs to be `gcp.projects.Service`.

@pgavlin Not sure by `.get` you mean `.getOutput` or similar method on `StackReference` because could not find any such method.

Any example code to handle such would be helpful.

**Comment by krishnamauryaext** [comment]
Another example is for exporting the output for enabled services in project. Enabling the services in some common pulumi project and then using that output in other projects to perform operation on them. The function expecting it to be `gcp.projects.Service` instead of `Output<T>`.

**Pull request merged**